staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go

```diff
@@ -60,8 +60,8 @@ type SpdyRoundTripper struct {
 	// conn is the underlying network connection to the remote server.
 	conn net.Conn
 
-	// Dialer is the dialer used to connect.  Used if non-nil.
-	Dialer *net.Dialer
+	// Dial specifies the dial function for creating TCP conntions.
+	Dial func(ctx context.Context, network, address string) (net.Conn, error)
```

Comment:
this is a breaking change

Reply:
Yes, maybe, change from net.Dialer to Dial function make it more flexible.

Reply:
Or maybe we can keep both

Reply:
just another thing you should take into considerations, maintainers are going to ask you for an use case of using a custom Dial function ;)

Reply:
Yes, I was making a spdy executor using this piece of code, and in my case we have to use a custom dial function.

Reply:
Yes, indeed this is a breaking change and in order to move forward, this PR needs a very strong justification.

Reply:
however, I think you can suggest depreccate the net.Dialer and add a new field to the struct similar to how net.http.Transport is implemented https://pkg.go.dev/net/http#Transport

```go
// DialContext specifies the dial function for creating unencrypted TCP connections.
	// If DialContext is nil (and the deprecated Dial below is also nil),
	// then the transport dials using package net.
	//
	// DialContext runs concurrently with calls to RoundTrip.
	// A RoundTrip call that initiates a dial may end up using
	// a connection dialed previously when the earlier connection
	// becomes idle before the later DialContext completes.
	DialContext func(ctx [context](https://pkg.go.dev/context).[Context](https://pkg.go.dev/context#Context), network, addr [string](https://pkg.go.dev/builtin#string)) ([net](https://pkg.go.dev/net).[Conn](https://pkg.go.dev/net#Conn), [error](https://pkg.go.dev/builtin#error))

	// DialTLSContext specifies an optional dial function for creating
	// TLS connections for non-proxied HTTPS requests.
	//
	// If DialTLSContext is nil (and the deprecated DialTLS below is also nil),
	// DialContext and TLSClientConfig are used.
	//
	// If DialTLSContext is set, the Dial and DialContext hooks are not used for HTTPS
	// requests and the TLSClientConfig and TLSHandshakeTimeout
	// are ignored. The returned net.Conn is assumed to already be
	// past the TLS handshake.
	DialTLSContext func(ctx [context](https://pkg.go.dev/context).[Context](https://pkg.go.dev/context#Context), network, addr [string](https://pkg.go.dev/builtin#string)) ([net](https://pkg.go.dev/net).[Conn](https://pkg.go.dev/net#Conn), [error](https://pkg.go.dev/builtin#error))

```

Reply:
@aojea Thanks for your great advice. I will do this way.

Mainly I did this PR is that net.Dialer is a concrete struct and we only can set timeout, deadline to control its Dial function how to create connection. While the upstream such as k8s.io/client-go/tools/remotecommand NewSPDYExecutor() at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go#L73 which make a SpdyRoundTripper from *restclient.Config, and the config object contains a Dial function. 

So, if somebody sets dial function in his/her *restclient.Config to make a SpdyRoundTripper object, he/she will expect the Dial function act, not lost. 

This is what I did last week. I dig into the code and find struct SpdyRoundTripper only contains a net.Dialer object, we cannot make a net.Dialer from Dial function to make the net.Dialer act as the dial function does.

After all, I know this is a breaking change and for backwards compatible, deprecating the net.Dialer is a great advice.

Thanks again, please let me know if lost some key point.

Reply:
Done, PTAL when you are free. @aojea @ardaguclu

Reply:
1. we need to have approval for owners of this area to move forward, I think that deprecating net.Dialer by a DialContext() function may make sense, but let the owners decide.
2. If you get approval to move forward, then the new parameter has to be backward compatible and explain how the new parameters are supposed to work, that you should explain what happen if net.Dialer and the new DialContext are != nil , ...
3. tests, there has to be unit tests asserting that is backward compatible and that the new functionality